apps/prairielearn/src/question-servers/freeform.js

```diff
@@ -1715,6 +1717,10 @@ export async function grade(submission, variant, question, question_course) {
 
     const context = await getContext(question, question_course);
     let data = {
+      // Note that `params` and `true_answer` can change during `parse()`/`grade()`.
+      // We can safely use the variant's values because we currently always write the
+      // updated values to both the submission and the variant. If that ever changes,
+      // we may need to update this to pull those values from the submission.
```

Comment:
FWIW, an alternative implementation here would have been to always pull the values from the submission. But regardless it's objectively more correct to pass in the updated `variant` when grading, so I'm sticking with that.